pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/buffer/impl/InMemTransactionBuffer.java

```diff
@@ -307,6 +307,7 @@ public CompletableFuture<Void> commitTxn(TxnID txnID, long lowWaterMark) {
                 txnBuffer.commitAt(committedAtLedgerId, committedAtEntryId);
                 addTxnToTxnIdex(txnID, committedAtLedgerId);
             }
+            updateLastDispatchablePosition(null);
```

Comment:
please see the changes carefully.

Reply:
I just promote a question here, right? There is no point in me arguing with you just to waste my time. What we do together in our spare time is make Pulsar better.

Reply:
You record a maximum stable position that can be used for consumption, but you set this position to null when a transaction is committed. This behavior puzzles me.
Because transaction committed will not change the previously recorded stable consumable position into a non-consumable position.

Reply:
Especially for  `InMemTransactionBuffer.java.`

Reply:
I introduced a filed `lastDispatchablePosition` to store the last position which is valid to dispatch, in `PersistentTopic`. And before return `lastDispatchablePosition`, check if `lastDispatchablePosition` is null, otherwise, search the position in BK.
About why set `lastDispatchablePosition` to null when commit txn, it is because of it's hard to determine which position to set, so, set to null to make it fallback to search the position from BK.

> Because transaction committed will not change the previously recorded stable consumable position into a
> non-consumable position.

I don't agree with this point, after commit a txn, the `lastDispatchablePosition` may have a chance to move forward, right? But, we don't know which position to set, and, even though we record the last entry position of the txn, what if non-txn messages published? 
Say, `txn1` contains `1, 2, 3` messages, and we record all the positions. Before commit `txn1`, there is a new non-txn message `4` published, when commit `txn1`, what value can we set to `lastDispatchablePosition`? So I set `lastDispatchablePosition` to null, we can get the `lastDispatchablePosition` by searching BK.

My  point is, the situations here is very complex, right? After commit a txn, the `lastDispatchablePosition` can be moved forward, or not. A topic contains `txn messages`, `non-txn messages`, `Marker`; `txn messages` contains `abort txn`, `ongoing txn`, `commit txn`. 
Set `lastDispatchablePosition` to null after commit/abort txn is the easiest way to make it consistent,

Reply:
Ok. More understand achieved.

Reply:
>About why set lastDispatchablePosition to null when commit txn, it is because of it's hard to determine which position to set, so, set to null to make it fallback to search the position from BK.

This is a good way for cache lastDispatchablePosition. There are some ways to do that. 
 1. your way
 2. When committing the transaction, keep the previous position because it is still the correct position. Then start a task to update, or mark it as needing to be updated from BK. Then the next time it is read, return to the previous position that is somewhat lagging but still valid. Then read from BK.

Reply:
Another point is that the transaction in `InMemTransactionBuffer ` does not effect the `lastDispatchablePosition`.

Reply:
> Another point is that the transaction in InMemTransactionBuffer  does not effect the lastDispatchablePosition.

It is because of I don't want to introduce `Separated` solution to `InMemoryTxnBuffer`, just make it consistent with `TopicTxnBuffer`, let it easier to maintain.

Reply:
You should do nothing here.

Reply:
OK, resolved